### ISSUE TITLE
Inputs.redfish says it reports fans. That does not seem to be the case with ILO4

### ISSUE THREAD

**Issue by bunnyevans** [influxdata/telegraf#14488]
### Relevant telegraf.conf

```toml
[[inputs.redfish]]
username = "redfish"
password = "REDfish!"
address = "https://tm200-ilo"

include_metrics = ["power", "thermal"]
insecure_skip_verify = true
```

### Logs from Telegraf

```text
2023-12-21T08:55:07Z I! Loading config: /etc/telegraf/telegraf.conf
2023-12-21T08:55:07Z I! Loading config: /etc/telegraf/telegraf.d/redfish.conf
2023-12-21T08:55:07Z E! Unable to open /var/log/telegraf/telegraf.log (open /var/log/telegraf/telegraf.log: permission denied), using stderr
2023-12-21T08:55:07Z I! Starting Telegraf 1.29.1 brought to you by InfluxData the makers of InfluxDB
2023-12-21T08:55:07Z I! Available plugins: 241 inputs, 9 aggregators, 30 processors, 24 parsers, 60 outputs, 6 secret-stores
2023-12-21T08:55:07Z I! Loaded inputs: redfish
2023-12-21T08:55:07Z I! Loaded aggregators:
2023-12-21T08:55:07Z I! Loaded processors:
2023-12-21T08:55:07Z I! Loaded secretstores:
2023-12-21T08:55:07Z W! Outputs are not used in testing mode!
2023-12-21T08:55:07Z I! Tags enabled: host=waffle
2023-12-21T08:55:07Z D! [agent] Initializing plugins
2023-12-21T08:55:07Z D! [agent] Starting service inputs
> redfish_power_powercontrol,address=tm200-ilo,host=waffle,source=waffle average_consumed_watts=0,interval_in_min=20i,max_consumed_watts=0,min_consumed_watts=0,power_capacity_watts=0,power_consumed_watts=0 1703148910000000000
> redfish_power_powersupplies,address=tm200-ilo,health=OK,host=waffle,name=HpServerPowerSupply,source=waffle,state=Enabled last_power_output_watts=0,line_input_voltage=0,power_capacity_watts=0 1703148910000000000
2023-12-21T08:55:11Z D! [agent] Stopping service inputs
2023-12-21T08:55:11Z D! [agent] Input channel closed
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=01-Inlet\ Ambient,source=waffle,state=Enabled reading_celsius=22,upper_threshold_critical=42,upper_threshold_fatal=46 1703148911000000000
2023-12-21T08:55:11Z D! [agent] Stopped Successfully
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=02-CPU,source=waffle,state=Enabled reading_celsius=40,upper_threshold_critical=70,upper_threshold_fatal=0 1703148911000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=04-P1\ DIMM\ 1-2,source=waffle,state=Enabled reading_celsius=48,upper_threshold_critical=89,upper_threshold_fatal=0 1703148911000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=05-VR\ P1,source=waffle,state=Enabled reading_celsius=62,upper_threshold_critical=115,upper_threshold_fatal=120 1703148911000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=06-VR\ MEM,source=waffle,state=Enabled reading_celsius=46,upper_threshold_critical=115,upper_threshold_fatal=120 1703148911000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=09-Side\ Ambient,source=waffle,state=Enabled reading_celsius=30,upper_threshold_critical=0,upper_threshold_fatal=0 1703148911000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=10-HDD\ MB\ Zone,source=waffle,state=Enabled reading_celsius=50,upper_threshold_critical=60,upper_threshold_fatal=0 1703148911000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=11-SSD\ MB\ zone\ 1,source=waffle,state=Enabled reading_celsius=45,upper_threshold_critical=60,upper_threshold_fatal=0 1703148911000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=12-Front\ Ambient,source=waffle,state=Enabled reading_celsius=29,upper_threshold_critical=0,upper_threshold_fatal=0 1703148911000000000
> redfish_thermal_temperatures,address=tm200-ilo,host=waffle,name=13-HDD\ ex-board,source=waffle,state=Absent reading_celsius=0,upper_threshold_critical=0,upper_threshold_fatal=0 1703148911000000000
> redfish_thermal_temperatures,address=tm200-ilo,host=waffle,name=14-PCI\ zone,source=waffle,state=Absent reading_celsius=0,upper_threshold_critical=0,upper_threshold_fatal=0 1703148911000000000
```

### System info

Latest Telegraf on Debian 12. HPE tm200, latest ILO

### Steps to reproduce

1.  have an HPE TM200 (gen9)
2. have an HPE Microserver (gen8)
3. telegraf --test --debug --input-filter redfish

### Expected behavior

Expect to see a 
redfish_thermal_fans
like the documentation mentions. 
This works on ILO5. (gen10)

### Actual behavior

Everything but that.
The same configuration pointed at ILO5 does produce fans.

### Additional info

ALL generated from:
curl https://ILO/redfish/v1/Chassis/1/Thermal/ -i --insecure -L -u UUU:PPPP 

ILO4 gen8 FANS snippet

```json
{
    "@odata.context": "/redfish/v1/$metadata#Chassis/Members/1/Thermal$entity",
    "@odata.id": "/redfish/v1/Chassis/1/Thermal/",
    "@odata.type": "#Thermal.1.2.0.Thermal",
    "Fans": [
        {
            "CurrentReading": 17,
            "FanName": "Fan 1",
            "Oem": {
                "Hp": {
                    "@odata.type": "#HpServerFan.1.0.0.HpServerFan",
                    "Location": "System",
                    "Type": "HpServerFan.1.0.0"
                }
            },
            "Status": {
                "Health": "OK",
                "State": "Enabled"
            },
            "Units": "Percent"
        }
    ],
```

ILO4 gen9 FANS snippet

```json
{
    "@odata.context": "/redfish/v1/$metadata#Chassis/Members/1/Thermal$entity",
    "@odata.id": "/redfish/v1/Chassis/1/Thermal/",
    "@odata.type": "#Thermal.1.2.0.Thermal",
    "Fans": [
        {
            "CurrentReading": 15,
            "FanName": "Fan 1",
            "Oem": {
                "Hp": {
                    "@odata.type": "#HpServerFan.1.0.0.HpServerFan",
                    "Location": "System",
                    "Type": "HpServerFan.1.0.0"
                }
            },
            "Status": {
                "Health": "OK",
                "State": "Enabled"
            },
            "Units": "Percent"
        }
    ],
```

ILO5 gen10 FANS snippet

```json
{
    "@odata.context": "/redfish/v1/$metadata#Thermal.Thermal",
    "@odata.etag": "W/\"0D69C1F2\"",
    "@odata.id": "/redfish/v1/Chassis/1/Thermal/",
    "@odata.type": "#Thermal.v1_7_1.Thermal",
    "Fans": [
        {
            "@odata.id": "/redfish/v1/Chassis/1/Thermal/#Fans/0",
            "MemberId": "0",
            "Name": "Fan 1",
            "Oem": {
                "Hpe": {
                    "@odata.context": "/redfish/v1/$metadata#HpeServerFan.HpeServerFan",
                    "@odata.type": "#HpeServerFan.v2_0_0.HpeServerFan",
                    "HotPluggable": false,
                    "Location": "System",
                    "Redundant": false
                }
            },
            "Reading": 11,
            "ReadingUnits": "Percent",
            "Status": {
                "Health": "OK",
                "State": "Enabled"
            }
        },
        {
            "@odata.id": "/redfish/v1/Chassis/1/Thermal/#Fans/1",
            "MemberId": "1",
            "Name": "Fan 2",
            "Oem": {
                "Hpe": {
                    "@odata.context": "/redfish/v1/$metadata#HpeServerFan.HpeServerFan",
                    "@odata.type": "#HpeServerFan.v2_0_0.HpeServerFan",
                    "HotPluggable": false,
                    "Location": "System",
                    "Redundant": false
                }
            },
            "Reading": 6,
            "ReadingUnits": "Percent",
            "Status": {
                "Health": "OK",
                "State": "Enabled"
            }
        },
        {
            "@odata.id": "/redfish/v1/Chassis/1/Thermal/#Fans/2",
            "MemberId": "2",
            "Name": "Fan 3",
            "Oem": {
                "Hpe": {
                    "@odata.context": "/redfish/v1/$metadata#HpeServerFan.HpeServerFan",
                    "@odata.type": "#HpeServerFan.v2_0_0.HpeServerFan",
                    "HotPluggable": false,
                    "Location": "System",
                    "Redundant": false
                }
            },
            "Reading": 6,
            "ReadingUnits": "Percent",
            "Status": {
                "Health": "OK",
                "State": "Enabled"
            }
        }
    ],
```

**Comment by powersj** [comment]
@bunnyevans thanks for the JSON output, it looks like we are missing two fields.
Can you please try the artifacts in:

https://github.com/influxdata/telegraf/pull/14492

**Comment by bunnyevans** [comment]
Assuming you mean, go there, download that and run it against the same configuration. 

Sadly, no change detected.

Here is the full output of that executable against the tm200
`
2023-12-22T11:10:26Z I! Loading config: /etc/telegraf/telegraf.conf
2023-12-22T11:10:26Z I! Loading config: /etc/telegraf/telegraf.d/redfish.conf
2023-12-22T11:10:26Z E! Unable to open /var/log/telegraf/telegraf.log (open /var/log/telegraf/telegraf.log: permission denied), using stderr
2023-12-22T11:10:26Z I! Starting Telegraf 1.30.0-958ed419 brought to you by InfluxData the makers of InfluxDB
2023-12-22T11:10:26Z I! Available plugins: 241 inputs, 9 aggregators, 30 processors, 24 parsers, 60 outputs, 6 secret-stores
2023-12-22T11:10:26Z I! Loaded inputs: redfish
2023-12-22T11:10:26Z I! Loaded aggregators:
2023-12-22T11:10:26Z I! Loaded processors:
2023-12-22T11:10:26Z I! Loaded secretstores:
2023-12-22T11:10:26Z W! Outputs are not used in testing mode!
2023-12-22T11:10:26Z I! Tags enabled: host=waffle
2023-12-22T11:10:26Z D! [agent] Initializing plugins
2023-12-22T11:10:26Z D! [agent] Starting service inputs
> redfish_power_powercontrol,address=tm200-ilo,host=waffle,source=waffle average_consumed_watts=0,interval_in_min=20i,max_consumed_watts=0,min_consumed_watts=0,power_capacity_watts=0,power_consumed_watts=0 1703243430000000000
> redfish_power_powersupplies,address=tm200-ilo,health=OK,host=waffle,name=HpServerPowerSupply,source=waffle,state=Enabled last_power_output_watts=0,line_input_voltage=0,power_capacity_watts=0 1703243430000000000
2023-12-22T11:10:30Z D! [agent] Stopping service inputs
2023-12-22T11:10:30Z D! [agent] Input channel closed
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=01-Inlet\ Ambient,source=waffle,state=Enabled reading_celsius=20,upper_threshold_critical=42,upper_threshold_fatal=46 1703243431000000000
2023-12-22T11:10:30Z D! [agent] Stopped Successfully
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=02-CPU,source=waffle,state=Enabled reading_celsius=40,upper_threshold_critical=70,upper_threshold_fatal=0 1703243431000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=04-P1\ DIMM\ 1-2,source=waffle,state=Enabled reading_celsius=47,upper_threshold_critical=89,upper_threshold_fatal=0 1703243431000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=05-VR\ P1,source=waffle,state=Enabled reading_celsius=62,upper_threshold_critical=115,upper_threshold_fatal=120 1703243431000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=06-VR\ MEM,source=waffle,state=Enabled reading_celsius=46,upper_threshold_critical=115,upper_threshold_fatal=120 1703243431000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=09-Side\ Ambient,source=waffle,state=Enabled reading_celsius=29,upper_threshold_critical=0,upper_threshold_fatal=0 1703243431000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=10-HDD\ MB\ Zone,source=waffle,state=Enabled reading_celsius=49,upper_threshold_critical=60,upper_threshold_fatal=0 1703243431000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=11-SSD\ MB\ zone\ 1,source=waffle,state=Enabled reading_celsius=43,upper_threshold_critical=60,upper_threshold_fatal=0 1703243431000000000
> redfish_thermal_temperatures,address=tm200-ilo,health=OK,host=waffle,name=12-Front\ Ambient,source=waffle,state=Enabled reading_celsius=28,upper_threshold_critical=0,upper_threshold_fatal=0 1703243431000000000
> redfish_thermal_temperatures,address=tm200-ilo,host=waffle,name=13-HDD\ ex-board,source=waffle,state=Absent reading_celsius=0,upper_threshold_critical=0,upper_threshold_fatal=0 1703243431000000000
> redfish_thermal_temperatures,address=tm200-ilo,host=waffle,name=14-PCI\ zone,source=waffle,state=Absent reading_celsius=0,upper_threshold_critical=0,upper_threshold_fatal=0 1703243431000000000
`

**Comment by bunnyevans** [comment]
It should be noted that

ilo4 gen8 is an HP Microserver gen8, they are lovely cute cubes with ONE FAN and a standard PSU.

ilo4 gen9 is an HP TM200 which also has ONE FAN (40mm?) and an external power brick. 

ilo5 gen10 is an HP ML30 with dual power and the PCI fan too.

If needed, I can also test against a DL20 gen10.

**Comment by powersj** [comment]
I have updated the PR with some additional debugging information. Need to see if any fans are even found. Please try the new artifacts. Thanks

**Comment by powersj** [comment]
@bunnyevans were you able to get the additional info?

**Comment by powersj** [comment]
@bunnyevans if you are able to get the logs I would appreciate it and we can look at this again.

**Comment by ryankolk** [comment]
Hi! I ran into the same issue and did some digging, thought I'd share. It turns out the output of the ilo4 endpoints are sanitized as some properties are pre-Redfish ([https://servermanagementportal.ext.hpe.com/docs/redfishservices/ilos/ilo5/ilo5_adaptation/#odata-version-http-header-requirements](https://servermanagementportal.ext.hpe.com/docs/redfishservices/ilos/ilo5/ilo5_adaptation/#odata-version-http-header-requirements)). This is only done when the `OData-Version` header is set.

Snippet of the response with `OData-Version` header:
```json
"Fans": [
  {
    "FanName": "Fan 1",
    "Oem": {
      "Hp": {
        "@odata.type": "#HpServerFan.1.0.0.HpServerFan",
        "Location": "System"
      }
    },
    "Status": {
    "Health": "OK",
    "State": "Enabled"
  }
},
...
]
```

And without header:
```json
"Fans": [
  {
    "CurrentReading": 7,
    "FanName": "Fan 1",
    "Oem": {
      "Hp": {
        "@odata.type": "#HpServerFan.1.0.0.HpServerFan",
        "Location": "System",
        "Type": "HpServerFan.1.0.0"
      }
    },
    "Status": {
        "Health": "OK",
        "State": "Enabled"
    },
    "Units": "Percent"
  },
...
]
```

The logs:
```
https://10.50.0.15/redfish/v1/Chassis/1/Thermal/
...
found 7 fans
{  Fan 1 <nil> <nil> <nil> <nil> <nil> <nil> <nil> {Enabled OK}}
{  Fan 2 <nil> <nil> <nil> <nil> <nil> <nil> <nil> {Enabled OK}}
{  Fan 3 <nil> <nil> <nil> <nil> <nil> <nil> <nil> {Enabled OK}}
{  Fan 4 <nil> <nil> <nil> <nil> <nil> <nil> <nil> {Enabled OK}}
{  Fan 5 <nil> <nil> <nil> <nil> <nil> <nil> <nil> {Enabled OK}}
{  Fan 6 <nil> <nil> <nil> <nil> <nil> <nil> <nil> {Enabled OK}}
{  Fan 7 <nil> <nil> <nil> <nil> <nil> <nil> <nil> {Enabled OK}}
```

With a bodged fix I was able to get the readings by setting the header based on endpoint, but it's not something you'd want in the codebase. but for completeness sake:
```go
//Replaced req.Header in getData
// if address does not end in Thermal/ set header
if(address[len(address)-8:] != "Thermal/"){
  req.Header.Set("OData-Version", "4.0")
}
```

I don't know what this means for ilo4 redfish support, but I'd be happy to help in any way I can!

**Comment by powersj** [comment]
@ryankolk,

Thanks for the info and offer to help.

> With a bodged fix I was able to get the readings by setting the header based on endpoint

Ah I did not realize we were setting this header per https://github.com/influxdata/telegraf/issues/8093. Seems this may be an iLO4-ism that is required for some of the endpoints, but not for the thermal one?

Is my understanding accurate?

Thanks!

**Comment by ryankolk** [comment]
Yeah that sounds about right! I believe the ILO4 was released before redfish, and some of the properties didn't match up when porting from the original rest structure to redfish. Instead of renaming some of the fields, they filtered some of the response. It feels like they forgot to rename these fields, but I can't know for sure.

As for #8093, the `@odata.id` property which is used to set the address for finding the Chassis, was originally called `href`, causing the getData function to try and load `https://ILO/` which returns an html page (with the first character being of course <).

Long story short, the ilo4 implementation of redfish feels a bit messy. Setting the header makes the response redfish compliant, but causes some properties to be discarded. The only reason the bodged fix works is because thermal isn't used to navigate to another address, and the structure is very similar.

What do you think would be a solution? Maybe something like adding a parameter to the config that disables the OData header on the Thermal endpoint?

**Comment by powersj** [comment]
Thanks for the additional background! That helps explain why iLO4 is in the weird state, while trying to support redfish.

I have updated #14659 to add a new parameter `workarounds` where a user can set it to `workarounds = ["ilo4-thermal"]` to unset the `OData-Version` header for the thermal endpoint. Would you be willing to give the artifacts from that PR a try?

Thanks!

**Comment by powersj** [comment]
Thank you very much for confirming.

> I'd recommend just comparing /Thermal.

Thanks for the feedback, I'll make that change and get this reviewed. Thanks again!

**Comment by ryankolk** [comment]
Thank you for your time and effort!
